### ISSUE TITLE
Add support for pushing stacked reviews to GitHub, GitLab, etc.

### ISSUE THREAD

**Issue by martinvonz** [martinvonz/jj#485]
## Description

For users who prefer one review per commit (as we do in this project), it would be useful to have some support for pushing a stack of commits and have it automatically create (or reuse) PRs for each commit. In the case of GitHub, we should also set the base branch appropriately then. I don't know about GitLab or any other providers we would care about.

Maybe we'd want commands like `jj github push` and `jj gitlab push` or something like, to make it clear that they're specifically for those providers.

**Comment by bklebe** [comment]
I love this feature in [Sapling](https://sapling-scm.com/docs/git/sapling-stack) and would love to see it in Jujutsu as well!

**Comment by FirelightFlagboy** [comment]
[`git town propose`](https://www.git-town.com/commands/propose) have a similar feature where it would create a new pull request for the current branch and setting the base as the configured parent branch.

**Comment by steveklabnik** [comment]
See also https://github.com/ejoffe/spr

and @sunshowers's fork, which has experimental jj support https://github.com/sunshowers/spr

**Comment by sunshowers** [comment]
Note that spr (and all other stacked diff tooling for GH, afaik) requires that you have push access to the repo you're proposing changes to.

**Comment by eopb** [comment]
For anyone who stumbles across this issue, I have a little nushell script that gets me some of the way there.

https://github.com/eopb/jj-gh-pr.nu

Even if you don't use nushell, maybe it'll still be useful to get ideas to adapt for your own scripts that you can use until `jj` has more built in support

**Comment by riwsky** [comment]
in the spirit of @eopb 's comment, here's some bash for submitting stacked PRs to graphite.dev from a jujutsu repo: https://gist.github.com/riwsky/38b17ea3fca70acf20a18c752663aff4

**Comment by PhilipMetzger** [comment]
There now are separate issues for both the GitHub (#4555) and Gerrit Integration (#4387). In my opinion this should be kept as a Umbrella task for more integrations.

**Comment by arxanas** [comment]
[topic: one command to handle all forges? replying to @dbarnett's comment in this thread since this is the forge-agnostic thread]

> @dbarnett: Yeah, tbh all the integrations I can think of for GitHub would have direct analogs for other forges, and I'm questioning whether it needs anything like an explicit jj github command at all. Besides some slight [terminology differences](https://about.gitlab.com/blog/2017/09/11/comparing-confusing-terms-in-github-bitbucket-and-gitlab/), if you can do it for GH you should probably be able to do it for any other supported remote.

I'm no longer in favor —

- git-branchless originally took the same approach and put every forge behind a [common `git submit` command](https://github.com/arxanas/git-branchless/wiki/Command:-git-submit).
- But I ended up agreeing with @martinvonz's points [in this Discord thread](https://discord.com/channels/968932220549103686/969291218347524238/1128718254576255076) that there are enough differences between forges that it's worth separating them.

Reproducing some messages from the Discord thread here:

> @martinvonz: i still feel like a generic command for sending for review or for "merging" a review will be hard to do with a good UX for the CLI. Specifically, it seems like the commands will have to have the union of all options needed by different forges (but probably not Google's internal one, and not including other internal ones)
>
> [...]
>
> @arxanas: How often do people use command-line options to affect the code review? I think most of mine would be either metadata in e.g. the commit message or actions in the web UI
> [...]
>
> @PhilipMetzger: iirc, the internal tool has submit and approve, which I both like
>
> @martinvonz: some frequently used options we have with our hg-based solution include for letting the PR get automatically rebased before getting landed, for publishing draft comments (previously entered in the web ui), and for specifying associated bugs [...] you can add additional bug associations when you send for review

---

[topic: verb choice]

- git-branchless uses `submit` to create a code review object,
- but Google uses "submit" for merging the actual code review:

> @martinvonz: we have hg mail for sending for review and hg submit for "merging"

- I do think "submit" is a little ambiguous and not a great term overall.
- `mail` was discussed as a reasonably forge-agnostic option.
  - It doesn't conflict with most forge concepts/terms.
  - I personally think it's a bit unintuitive/undiscoverable.
- I think `upload` was discussed as well.
  - Fairly descriptive in my opinion, although it's a little generic.
    - I wonder if we did user research whether it would be found to be most intuitive?
  - Suggests that there would be a corresponding `download` command.
    - To be fair, "downloading"/getting a local copy of a remote code review is actually a reasonably common workflow.
- We could reuse `push` as the "standard" subcommand for this purpose.
  - If this term were adopted:
    - `jj git push` would create a remote branch.
      - git-branchless calls this case the "branch" forge in its `git submit`.
    - `jj github push`/`jj gitlab push` would actually create pull requests/merge requests respectively.
  - It reuses existing terminology in a reasonably intuitive way.
    - Can reuse `pull` verb as well... although maybe the semantics of three-way merge mean that we _shouldn't_ use push/pull as the terms?
    - Maybe three-way-merge-ness is a property of the forge? Only the "branch" forge would use it?
      - Maybe that's the difference between `push`/`upload` and `pull`/`download`?
    - Okay, so maybe this term is not that intuitive?

---

[topic: standardization and code sharing]

In git-branchless, I haven't been able to abstract over the different forges that much. The current common interface is here:

https://github.com/arxanas/git-branchless/blob/db7cc710658190f83f887adb8c825aaaef014b34/git-branchless-submit/src/lib.rs#L158-L180

In particular, I haven't been able to abstract over topologically submitting updates and setting dependencies, although it might be possible.
- For efficiency, you often want to batch creates/updates rather than issue 1x serialized command per create/update.
  - But then each forge needs to know how to topologically sort within a batch, and when it's appropriate/necessary to do so?
- See https://github.com/arxanas/git-branchless/discussions/1259 for discussion of certain implementation issues for the GitHub forge.
  - It also motivates a separate "land"/"submit"/"merge"/whatever subcommand, separate from the subcommand that only creates/updates the code review objects.
- There's a common problem where, when you merge a commit at the bottom of a stack, then rebase dependent commits + code reviews, the newly-rebased code reviews "forget" that they were dependent on the now-merged code reviews.
  - Not sure if we can or want to solve it generically. It probably requires extra storage; both in-band and out-of-band are options.
  - This arose for at least Phabricator and GitHub so far, and I'm sure would continue to arise for future forges.

Ideally we would also reuse certain UI components to make the output formatting consistent.
- I had been iterating on that for git-branchless.
- Originally I printed one line per status ("updated"/"already up to date"/"error", etc.).
- But i think the best UI is one line per commit, with each line including an identifier and a link to to the remote object, if any.
- Users likely would want to template the output for use in their own scripting.

**Comment by aaomidi** [comment]
Another possible integration is with aviator.co and their stacked PRs

**Comment by pelme** [comment]
To play with this idea, I created a Python script that creates a stack of PRs from jj changes: https://gist.github.com/pelme/ebd043a6a0487511ebfd804b5976e475. Have been using it the last couple of days for all my github PR needs!

**Comment by matklad** [comment]
Related: Sapling has ReviewStack (https://sapling-scm.com/docs/addons/reviewstack/), an alternative front-end for GitHub pull-review interface which:

* supports stacked diffs
* but otherwise has look&feel of the normal GitHub review

The TL;DR is that you replace github.com in `https://github.com/tigerbeetle/tigerbeetle/pull/2360` with `https://reviewstack.dev/` and get a nice web UI for reviews. 

If `jj` could make use of that, that'd be cool!

**Comment by martinvonz** [comment]
I think I've heard that Review stack is unmaintained. Not sure I remember that correctly. But we should check that before we add support for it.

**Comment by xgwang-scienceos** [comment]
@martinvonz At this point, are `jj mail` and `jj submit` the chosen commands for managing stacked PRs? Or are we still in the process of deciding the verbs?

Thanks!

**Comment by martinvonz** [comment]
Those are the verbs we use at Google. I think people suggested `jj gerrit send` for sending something for review with Gerrit. I can't remember what the current consensus is for creating GitHub PRs and GitLab MRs. I don't think we've decided to standardize on the same verbs across forges anyway.

**Comment by xgwang-scienceos** [comment]
Thanks for the quick response. Got it. In the case of multi forges support, it does make sense for the forge become the subcommand, each forge use their own familiar verbs. IMHO the best/fastest way to align is to respect original designs and provide features for customizable alignment. The jj alias feature does that very well. A team can easily create an team-wide alias to use verbs that makes sense for them. 

Let's say the general interface is `jj <forge> <forge-verb>`. Is the following a good set of steps for implementing stacked reviews?
- Step1: Add jj stacked review interfaces
- Step2: Implement stack representation in jj
- Step3: Implement jj stacked review interfaces
- Step4: Integrate forge review APIs

I'm asking this because our team already heavily rely on `jj`. The only missing feature to achieve Google `hg` like code review experience is stacking. I'd like to see if our team can implement stacking in jj to unblock ourself. 

Thank you!

**Comment by martinvonz** [comment]
I'm not sure we need any interfaces for stacked review. I think supports for stacked reviews is mostly limited by the forge you want to use. Which forge do you plan to use?

**Comment by xgwang-scienceos** [comment]
> I'm not sure we need any interfaces for stacked review. I think supports for stacked reviews is mostly limited by the forge you want to use. Which forge do you plan to use?

We plan to use Gerrit

**Comment by xgwang-scienceos** [comment]
At the moment, is this still the suggested way for Gerrit integration? https://jj-vcs.github.io/jj/latest/FAQ/#how-do-i-integrate-jujutsu-with-gerrit

**Comment by omarkohl** [comment]
EDIT: I ended up converting the script into a proper tool, named `jip`, that I am very happy with. https://github.com/omarkohl/jip

## Features

- **One commit == one PR** — each commit gets its own focused, reviewable PR,
  with title and description taken from the commit message
- **1-step PR creation** — By default you need to create a bookmark (branch),
  push it and finally open the PR with title and description. `jip send` does
  all that in one command.
- **Diff comments on update** — when you update a PR, jip posts a diff
  showing only what you changed (rebase noise is filtered out)
- **No local state** — all information lives in the PR itself on GitHub;
  reviewers and contributors don't need jip, and teams can freely mix jip and
  non-jip users
- **Automatic bookmarks** — no need to manually create jj bookmarks
- **Supports forks** — works both for repositories where you have write access as well as forked ones
- **Normal GitHub merge** — no special "land" command, just merge PRs as usual
- **Stacks may contain merge commits** — supports non-linear revision graphs

----

**Original comment:**

I also wrote a [Python script](https://github.com/omarkohl/dotfiles/blob/main/bin/stackpr) to do this for jj with GitHub and I use it frequently. It adds (and updates on every execution) PR descriptions like the following.

<img width="826" height="382" alt="Image" src="https://github.com/user-attachments/assets/f39eedb7-85e6-49c5-9080-ccf2bc643241" />

**Comment by doy-materialize** [comment]
it would be nice to get support for https://github.github.com/gh-stack/ once it is fully released!

**Comment by BatmanAoD** [comment]
Similarly, GitLab has an experimental stacked diffs feature: https://docs.gitlab.com/user/project/merge_requests/stacked_diffs/

Edit: never mind; see comment below.

**Comment by mxey** [comment]
> Similarly, GitLab has an experimental stacked diffs feature: https://docs.gitlab.com/user/project/merge_requests/stacked_diffs/

This is primarily a feature in the Gitlab CLI. The only server-side feature AFAIK is support for MR dependencies in Gitlab Ultimate.

**Comment by fynnsu** [comment]
From the FAQ for Github Stacked PRs:
https://github.github.com/gh-stack/faq/#will-this-work-with-a-different-tool-for-stacking

> 
> ## Will this work with a different tool for stacking?
> 
> Yes, you can continue to use your tool of choice (e.g., jj, Sapling, ghstack, git-town, etc.) to manage stacks locally and push up your branches to GitHub.
> 
> Stacked PRs on GitHub are based on the standard pull request model — any tool that creates PRs with the correct base branches can work with them. The gh stack CLI is purpose-built for the GitHub experience, but other tools that manage branch chains should be compatible.
> 
> You can also use the GitHub CLI in conjunction with other tools to open your PRs as a stack:
> ```bash
> # Create a stack of branches locally using jj
> jj new main -m "first change"
> jj bookmark create change1 --revision @
> # ...
> 
> jj new -m "second change"
> jj bookmark create change2 --revision @
> # ...
> 
> jj new -m "third change"
> jj bookmark create change3 --revision @
> # ...
> 
> # Use gh stack to submit a stack of PRs
> gh stack init --adopt change1 change2 change3
> gh stack submit
> ```

**Comment by BatmanAoD** [comment]
@mxey Ah, thanks, I missed that there were no UI or API changes planned.

**Comment by jdve** [comment]
I'm creating stacked MRs in GitLab using this simple script:

```bash
#!/bin/bash

changes=$(jj log -r 'trunk():: & ~trunk()' --no-graph -T 'change_id.short() ++ "\n"' --reversed)

target="main"

for change in $changes; do
    jj git push -c "$change" \
        -o merge_request.create \
        -o merge_request.target="$target"

    target=$(jj log -r "$change" --no-graph -T 'bookmarks.first()')
done
```

This pushes each commit in `trunk():: & ~trunk()` to GitLab as an individual MR, making sure that each MR properly targets the previous bookmark, so that GitLab detects the MRs as a stack.
